### PR TITLE
[MLAAttention] Clear Cudagraph padded region of FI decode Attention kernel

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
## Purpose
We see flashinfer MLA kernel introducing NaNs in the padded cudagraph region. This PR zeros out the output of the `trtllm_batch_decode_with_kv_cache_mla` kernel. 

Additionally reverts PR #37442 to as we observe some OOMs. 
Fixes : #37777 

Flashinfer issue : https://github.com/flashinfer-ai/flashinfer/issues/2883

## Test Plan
Run `nvidia/DeepSeek-R1-0528-NVFP4-v2` in a wide-ep setup and run lm_evals. 

## Test Result
I see that lm_evals pass locally

### CODE REVIEW

File: vllm/v1/attention/backends/mla/flashinfer_mla.py [hunk 1]
```diff
@@ -235,14 +198,15 @@ def forward_mqa(
...
-            if num_real < o.shape[0]:
-                o[num_real:] = 0
+        # Flashinfer MLA kernels introduces NaNs in padded regions in
+        # some cases.  We need to zero out the padded regions to avoid
+        # NaNs in the output.
+        assert o.size(0) == attn_metadata.decode.seq_lens.size(0), (
+            f"output shape {o.size()} != "
+            f"seq_lens shape {attn_metadata.decode.seq_lens.size()}"
+        )
+        o[attn_metadata.decode.seq_lens == 0] = 0
```
Comment (LucasWilkinson) [inline vllm/v1/attention/backends/mla/flashinfer_mla.py]: nit: I think `nan_to_num` could be faster (single kernel vs 2) or we should precompute `attn_metadata.decode.seq_lens == 0`

Comment (LucasWilkinson) [reply vllm/v1/attention/backends/mla/flashinfer_mla.py]: Quick claude benchmark:

```
  ┌───────────────────────────┬─────────────────┬──────────────────┬─────────┐                                                                                                                                                 
  │           Shape           │ nan_to_num (µs) │ mask assign (µs) │ Speedup │                                                                                                                                                 
  ├───────────────────────────┼─────────────────┼──────────────────┼─────────┤                                                                                                                                                 
  │ (64, 128, 512, pad=8)     │            5.95 │            10.64 │   1.79x │
  ├───────────────────────────┼─────────────────┼──────────────────┼─────────┤                                                                                                                                                 
  │ (128, 128, 512, pad=16)   │            7.77 │            18.57 │   2.39x │                                                                                                                                                 
  ├───────────────────────────┼─────────────────┼──────────────────┼─────────┤                                                                                                                                                 
  │ (256, 128, 512, pad=32)   │           13.53 │            34.75 │   2.57x │                                                                                                                                                 
  ├───────────────────────────┼─────────────────┼──────────────────┼─────────┤                                                                                                                                                 
  │ (512, 128, 512, pad=64)   │           35.26 │            90.36 │   2.56x │
  ├───────────────────────────┼─────────────────┼──────────────────┼─────────┤                                                                                                                                                 
  │ (1024, 128, 512, pad=128) │           66.94 │           176.25 │   2.63x │
  ├───────────────────────────┼─────────────────┼──────────────────┼─────────┤                                                                                                                                                 
  │ (2048, 128, 512, pad=256) │          166.12 │           348.30 │   2.10x │
  └───────────────────────────┴─────────────────┴──────────────────┴─────────┘   
```

Comment (varun-sundar-rabindranath) [reply vllm/v1/attention/backends/mla/flashinfer_mla.py]: Just worried that nan_to_num on the whole tensor might mask future bugs where the kernel itself produces nans. Let me see if there is a better way that masks only the padded region (precomputing the padding mask, like you mentioned).

File: vllm/v1/attention/backends/mla/cutlass_mla.py [hunk 2]
```diff
@@ -223,15 +218,7 @@ def _sm100_cutlass_mla_decode(
...
-        # Reuse pre-allocated zero-init output buffer to avoid a memset
-        # kernel on every CUDA graph replay.
-        if (
-            self._decode_out is None
-            or self._decode_out.shape[0] < B_q
-            or self._decode_out.dtype != dtype
-        ):
-            self._decode_out = q_nope.new_zeros((B_q, MAX_HEADS, D_latent), dtype=dtype)
-        out = self._decode_out[:B_q]
+        out = q_nope.new_empty((B_q, MAX_HEADS, D_latent), dtype=dtype)
```
[the remark this answers is not in the document]
Comment (varun-sundar-rabindranath) [reply vllm/v1/attention/backends/mla/cutlass_mla.py:221]: this is just reverting PR 37442 . This should be fine .

File: vllm/v1/attention/backends/mla/flashinfer_mla.py (hunk 1 above)
Comment (elvircrn) [reply vllm/v1/attention/backends/mla/flashinfer_mla.py]: Thanks, good info.

Comment (varun-sundar-rabindranath) [reply vllm/v1/attention/backends/mla/flashinfer_mla.py]: focusing on padding <=16 (how the cudagraph sizes are computed) -- I have the following numbers, 
```
┌───────────────────────────┬─────────────────┬──────────────────┬───────────────────┐                                                                                                                                                                                                                                      
│           Shape           │ nan_to_num (µs) │ mask assign (µs) │ pre-mask assign (µs) │                                                                                                                                                                                                                                   
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (1, 128, 512, pad=0)      │            9.82 │            36.07 │             16.60 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (2, 128, 512, pad=1)      │           12.25 │            38.10 │             14.92 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (4, 128, 512, pad=1)      │           13.31 │            37.40 │             15.16 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (8, 128, 512, pad=2)      │           12.93 │            37.42 │             16.47 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (12, 128, 512, pad=4)     │           12.63 │            38.09 │             16.44 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (16, 128, 512, pad=8)     │           11.54 │            35.09 │             14.38 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (24, 128, 512, pad=12)    │           12.12 │            36.36 │             15.24 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (32, 128, 512, pad=16)    │           11.75 │            34.67 │             16.12 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (64, 128, 512, pad=16)    │           12.16 │            35.35 │             15.55 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (128, 128, 512, pad=16)   │            9.78 │            34.79 │             16.19 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (192, 128, 512, pad=16)   │           10.28 │            29.60 │             22.55 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (256, 128, 512, pad=16)   │           12.22 │            32.64 │             28.67 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (320, 128, 512, pad=16)   │           14.29 │            39.11 │             34.82 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (384, 128, 512, pad=16)   │           16.48 │            46.70 │             41.01 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (512, 128, 512, pad=16)   │           20.46 │            59.32 │             55.25 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (640, 128, 512, pad=16)   │           26.36 │            75.84 │             71.70 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (768, 128, 512, pad=16)   │           33.09 │           106.07 │            102.25 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (896, 128, 512, pad=16)   │           41.23 │           131.10 │            127.05 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤                                                                                                                                                                                                                                      
│ (1024, 128, 512, pad=16)  │           47.12 │           149.54 │            145.51 │                                                                                                                                                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┤ 
```

precomputing mask helps -- This is the current implementation.  however nan_to_num is the clear winner. 

I am not using nan_to_num for the reason mentioned above (in https://github.com/vllm-project/vllm/pull/37815#discussion_r2976425455) .. 
what do you guys think ? Thanks.

File: vllm/model_executor/layers/attention/mla_attention.py [hunk 3]
```diff
@@ -1484,6 +1486,11 @@ def __init__(
             vllm_config, self.model_config.dtype
         )
 
+        # GPU persistent mask for cudagraph padding sequences.
+        self.decode_pad_mask = torch.zeros(
+            scheduler_config.max_num_seqs, dtype=torch.bool, device=device
+        )
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/attention/mla_attention.py]: I looked around a bit and settled here for a place for `decode_pad_mask` persistent tensor.
happy to move it elsewhere if it is better. Please let me know 🙌

Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/attention/mla_attention.py]: no longer relevant.

File: vllm/v1/attention/backends/mla/flashinfer_mla.py (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/attention/backends/mla/flashinfer_mla.py]: I couldn't find a better way to just use torch operations for the desired effect. I added a zero out kernel instead. 
the following are the numbers running the kernels inside a cudagraph (so as to not include the triton launch overhead) --  

```
┌───────────────────────────┬─────────────────┬──────────────────┬───────────────────┬───────────────────────┐                                                                                                      
│           Shape           │ nan_to_num (µs) │ mask assign (µs) │ pre-mask assign (µs) │ triton zero rows (µs) │                                                                                                   
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (1, 128, 512, pad=0)      │            2.03 │             3.82 │              2.85 │                  1.67 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (2, 128, 512, pad=1)      │            2.81 │             5.04 │              3.66 │                  4.40 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (4, 128, 512, pad=1)      │            2.96 │             5.04 │              3.54 │                  4.23 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (8, 128, 512, pad=2)      │            2.84 │             5.01 │              3.69 │                  4.27 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (12, 128, 512, pad=4)     │            2.85 │             5.16 │              3.65 │                  4.38 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (16, 128, 512, pad=8)     │            3.01 │             5.47 │              4.02 │                  4.66 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (24, 128, 512, pad=12)    │            3.07 │             6.28 │              5.14 │                  4.62 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (32, 128, 512, pad=16)    │            3.38 │             7.06 │              5.99 │                  4.50 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (64, 128, 512, pad=16)    │            5.08 │            10.55 │              9.44 │                  4.64 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (128, 128, 512, pad=16)   │            8.02 │            21.09 │             19.78 │                  4.50 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (192, 128, 512, pad=16)   │           10.66 │            30.18 │             28.75 │                  4.44 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (256, 128, 512, pad=16)   │           13.41 │            39.01 │             37.71 │                  4.31 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (320, 128, 512, pad=16)   │           16.04 │            47.97 │             46.62 │                  4.43 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (384, 128, 512, pad=16)   │           18.57 │            56.96 │             55.38 │                  4.46 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (512, 128, 512, pad=16)   │           23.82 │            74.72 │             73.24 │                  4.19 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (640, 128, 512, pad=16)   │           29.21 │            92.60 │             91.06 │                  4.34 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (768, 128, 512, pad=16)   │           34.37 │           110.67 │            108.88 │                  4.45 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (896, 128, 512, pad=16)   │           39.54 │           128.35 │            126.90 │                  4.45 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤                                                                                                      
│ (1024, 128, 512, pad=16)  │           45.05 │           145.93 │            144.68 │                  4.47 │                                                                                                      
├───────────────────────────┼─────────────────┼──────────────────┼───────────────────┼───────────────────────┤ 
```

File: vllm/v1/attention/backends/mla/utils.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        col_offsets = tl.arange(0, BLOCK_SIZE)
+        out_ptrs = out_ptr + row * row_stride + col_offsets
+        for c in tl.range(0, tl.cdiv(num_cols, BLOCK_SIZE)):
+            mask = col_offsets + c * BLOCK_SIZE < num_cols
+            tl.store(
+                out_ptrs,
+                tl.zeros([BLOCK_SIZE], dtype=out_ptr.dtype.element_ty),
+                mask=mask,
+            )
+            out_ptrs += BLOCK_SIZE
```
Comment (varun-sundar-rabindranath) [inline vllm/v1/attention/backends/mla/utils.py:35]: @LucasWilkinson @elvircrn  the kernel can use a fresh pair of eyes. Thanks 🙌

Comment (elvircrn) [reply vllm/v1/attention/backends/mla/utils.py:35]: Lgtm!

**Review by MatthewBonanni — APPROVED** [review]
LGTM
